### PR TITLE
prov/efa: Fix efa_rx_pkts_held counter drift

### PR DESCRIPTION
**Pull request by gjamesli2126**
## Original Issue:
  efa_rdm_ep->efa_rx_pkts_held drifted with `FI_EFA_RX_COPY_UNEXP=0` / `FI_EFA_RX_COPY_OOO=0`.

## Fixed with: Track efa_rx_pkts_held with a per-pkt flag
- Funnel `held++/--` through `EFA_RDM_PKE_HELD_BY_PROGRESS` flag.
- Add missing `held++` in `pke_get_unexp()` (rx_copy_unexp=0).
- Add missing `held++` in `pke_get_ooo_pke()` (rx_copy_ooo=0).

## Fixed with: Release held rx pkt on local-read teardown
Local-read txe holds pkt only via `txe->local_read_pkt_entry`; if torn down before read completes, pkt was orphaned.
- `txe_release()`: release pkt if non-NULL.
- `txe_construct()`: NULL-init.
- `pke_handle_rma_read_completion()` success path: NULL the field to avoid double-free.

## Validation
Debug build on p5en with `FI_PROVIDER=efa FI_EFA_RX_COPY_UNEXP=0 FI_EFA_RX_COPY_OOO=0`.
Oracle: `assert(held + posted + to_post == rx_pool_size)` at `efa_rdm_ep_utils.c:963`.

| Test                         | Path            | Pre-fix   | Post-fix 
|------------------------------|-------------------|------------|---------
| `fi_unexpected_msg`             | unexpected-recv  | abort@963 | clean   
| `fi_rdm_tagged_bw`               | OOO + long-msg | abort@963 | clean   
| `fi_rdm_tagged_bw -D cuda`| CUDA local-read  | abort@963 | clean   

## Fix double-free in local-read unit test
                                                                     
  `txe_release` now frees the held pkt on error path, so the test must not release it again.

  - **Happy path:** NULL `txe->local_read_pkt_entry` after manual release; teardown's `txe_release` would otherwise double-free.
  - **Unhappy path:** drop the manual release entirely (`txe_release` already handles it).
  - **Both paths:** verify `efa_rx_pkts_held` / `efa_rx_pkts_to_post` deltas + held flag.

## Out of scope/TODO: queued blocking copy mid-batch leak
  `flush_queued_blocking_copy_to_hmem()` early-return leaks pkts of other still-alive rxes.
  Naive release double-frees with rxe error path.

### CODE REVIEW

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 1]
```diff
@@ -1770,9 +1770,10 @@ int efa_rdm_rxe_post_local_read_or_queue(struct efa_rdm_ope *rxe,
 	/* The rx pkts are held until the local read completes */
 	if (err)
 		efa_rdm_txe_release(txe);
-	else if (txe->local_read_pkt_entry->alloc_type == EFA_RDM_PKE_FROM_EFA_RX_POOL)
-		txe->ep->efa_rx_pkts_held++;
-
+	else{
```
Comment (sunkuamzn) [inline prov/efa/src/rdm/efa_rdm_ope.c]: nit: space
`else {`

Please also add braces to the if block

**Review by sunkuamzn — COMMENTED** [review]
Please check CI failures

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 1 above)
Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Also updated the unit test. All unit test & necessary fabtests passed.

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 2]
```diff
@@ -1770,9 +1770,10 @@ int efa_rdm_rxe_post_local_read_or_queue(struct efa_rdm_ope *rxe,
 	/* The rx pkts are held until the local read completes */
 	if (err)
```
Comment (sunkuamzn) [inline prov/efa/src/rdm/efa_rdm_ope.c]: please add braces to the if block so that we don't braces on just the else block. e.g.

```
if (err) {
    efa_rdm_txe_release(txe);
} else {
    /*Local RDMA read posted*/
    efa_rdm_pke_mark_held(txe->local_read_pkt_entry);
}
```

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 3]
```diff
@@ -151,6 +152,13 @@ void efa_rdm_txe_release(struct efa_rdm_ope *txe)
 		txe->internal_flags &= ~EFA_RDM_OPE_QUEUED_FLAGS;
 	}
 
+	/* Release held rx-pool pkt on early teardown; success path NULLs to avoid double-release.*/
+	if (txe->local_read_pkt_entry &&
+	    txe->local_read_pkt_entry->alloc_type == EFA_RDM_PKE_FROM_EFA_RX_POOL) {
```
Comment (sunkuamzn) [inline prov/efa/src/rdm/efa_rdm_ope.c]: Why only release when the pke is from `EFA_RDM_PKE_FROM_EFA_RX_POOL`?

Looking at https://github.com/ofiwg/libfabric/blob/820d49724cb19e4732579ba423ca991b841356d4/prov/efa/src/rdm/efa_rdm_ope.c#L1288, the pke can either be from the rx_pkt_pool or the readcopy_pkt_pool

If the pke is from readcopy_pkt_pool, we should still release it?

File: prov/efa/test/efa_unit_test_ope.c [hunk 4]
```diff
@@ -345,12 +347,28 @@ void test_efa_rdm_rxe_post_local_read_or_queue_impl(struct efa_resource *resourc
...
 
-	/* Clean up the rx entry no matter what returns */
-	efa_rdm_pke_release_rx(pkt_entry);
+	if (efa_rdm_pke_read_return == FI_SUCCESS) {
+		/* mark_held fired: held++, flag set */
+		struct efa_rdm_ope *txe;
+		assert_int_equal(efa_rdm_ep->efa_rx_pkts_held, held_before + 1);
+		assert_int_equal(efa_rdm_ep->efa_rx_pkts_to_post, to_post_before);
+		assert_true(pkt_entry->flags & EFA_RDM_PKE_HELD_BY_PROGRESS);
+		efa_rdm_pke_release_rx(pkt_entry);
```
Comment (sunkuamzn) [inline prov/efa/test/efa_unit_test_ope.c]: I would call `efa_rdm_pke_handle_rma_read_completion` here so that you can also test that `efa_rdm_pke_handle_rma_read_completion` sets `txe->local_read_pkt_entry = NULL`

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 2 above)
Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Thanks for the suggest. I missed that.

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 3 above)
Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Yes we should release it too!

File: prov/efa/src/rdm/efa_rdm_pke_nonreq.c [hunk 5]
```diff
@@ -512,7 +512,8 @@ void efa_rdm_pke_handle_rma_read_completion(struct efa_rdm_pke *context_pkt_entr
 			if (txe->peer == NULL) {
 				data_pkt_entry = txe->local_read_pkt_entry;
 				assert(data_pkt_entry->payload_size > 0);
-				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() below */
+				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() below; NULL the txe ref so efa_rdm_txe_release() doesn't double-free. */
+				txe->local_read_pkt_entry = NULL;
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_nonreq.c]: This should be moved after the pkt entry release, i.e. `efa_rdm_pke_handle_data_copied`

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 3 above)
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_ope.c]: If local read pkt entry is not finished and you release it here, will we get a use-after-free bug when the completion arrives later?

Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Good catch! I'll gate it with `if (txe->efa_outstanding_tx_ops==0)`

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Also, is it necessary to check whether the alloc type is rx pool? If it is allocated from the read copy pool, don't we want to release too ?

Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Thanks Shi, Sai pointed out this part too. I've made the change in my local.

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_ope.c]: As I can tell we always release pkt entries before releasing op entries. Then we should not worry about this local read entry if it is released earlier. Is there a scenario that we can release txe early? In the ep close's bufferpool destroy, we first release pkt entries (for debug build), then release op entries. For release build, that pkt list is not available so no pke release before ope release. But finally the bufferpool should be destroyed and memories are freed. So no leak. What is the value of such pkt entry in txe release?

Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: After a 2nd look, you're right. The `pkt_entry` is always released by the protocol completion handler (success) or the caller's error branch (failure), so `txe_release` shouldn't touch it. I'll fix it. Thanks!

File: prov/efa/src/rdm/efa_rdm_pke_nonreq.c [hunk 6]
```diff
@@ -512,9 +512,10 @@ void efa_rdm_pke_handle_rma_read_completion(struct efa_rdm_pke *context_pkt_entr
 			if (txe->peer == NULL) {
 				data_pkt_entry = txe->local_read_pkt_entry;
 				assert(data_pkt_entry->payload_size > 0);
-				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() below */
 				efa_rdm_tracepoint(rx_pke_local_read_copy_payload_end, (size_t) data_pkt_entry, data_pkt_entry->payload_size, data_pkt_entry->ope->msg_id, (size_t) data_pkt_entry->ope->cq_entry.op_context, data_pkt_entry->ope->total_len);
 				efa_rdm_pke_handle_data_copied(data_pkt_entry);
+				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() below; NULL the txe ref so efa_rdm_txe_release() doesn't double-free. */
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_nonreq.c]: `below` -> `above`, or you keep that original comment as is and only have your new comment here

**Review by shijin-aws — DISMISSED** [review]

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by gjamesli2126** [comment]
bot:aws:retest

**Comment by gjamesli2126** [comment]
bot:aws:retest

File: prov/efa/src/rdm/efa_rdm_pke_nonreq.c [hunk 7]
```diff
@@ -512,9 +512,10 @@ void efa_rdm_pke_handle_rma_read_completion(struct efa_rdm_pke *context_pkt_entr
 			if (txe->peer == NULL) {
 				data_pkt_entry = txe->local_read_pkt_entry;
 				assert(data_pkt_entry->payload_size > 0);
-				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() below */
 				efa_rdm_tracepoint(rx_pke_local_read_copy_payload_end, (size_t) data_pkt_entry, data_pkt_entry->payload_size, data_pkt_entry->ope->msg_id, (size_t) data_pkt_entry->ope->cq_entry.op_context, data_pkt_entry->ope->total_len);
 				efa_rdm_pke_handle_data_copied(data_pkt_entry);
+				/* Hand off pkt release to efa_rdm_pke_handle_data_copied() above; NULL the txe ref so efa_rdm_txe_release() doesn't double-free. */
```
Comment (sunkuamzn) [inline prov/efa/src/rdm/efa_rdm_pke_nonreq.c]: Is there still a double free?

Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_pke_nonreq.c]: No there is not. Will fix the comment. I've removed the commit msg accordingly.

**Review by shijin-aws — DISMISSED** [review]

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 8]
```diff
@@ -1768,11 +1768,12 @@ int efa_rdm_rxe_post_local_read_or_queue(struct efa_rdm_ope *rxe,
 	txe->internal_flags |= EFA_RDM_OPE_INTERNAL;
 	err = efa_rdm_ope_post_remote_read_or_queue(txe);
 	/* The rx pkts are held until the local read completes */
-	if (err)
+	if (err){
```
Comment (jiaxiyan) [inline prov/efa/src/rdm/efa_rdm_ope.c]: nit: space between `) {`

File: prov/efa/src/rdm/efa_rdm_pke.c [hunk 9]
```diff
@@ -253,6 +260,20 @@ void efa_rdm_pke_release_rx_list(struct efa_rdm_pke *pkt_entry)
 	}
 }
 
+/**
+   * @brief Mark an rx-pool pkt as held by progress; release_rx() will decrement.
```
Comment (jiaxiyan) [inline prov/efa/src/rdm/efa_rdm_pke.c]: nit: comment wasn't aligned. There is an extra space on each line.

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 8 above)
Comment (gjamesli2126) [reply prov/efa/src/rdm/efa_rdm_ope.c]: good catch, I missed it again. I should get this new habit.

**Review by sunkuamzn — APPROVED** [review]

**Comment by gjamesli2126** [comment]
bot:aws:retest
